### PR TITLE
Strategies: IEngine for context dependent ids.

### PR DESCRIPTION
**Pull request by sergey-shandar**
## Description

Strategies: IEngine for context dependent ids.

## Checklist

- [ ] I have read the [_Submitting Changes_](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#making-changes) section of [`CONTRIBUTING.md`](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md)
- [ ] The title of the PR is clear and informative
- [ ] The appropriate [change log has been updated](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#updating-the-change-log)
- [ ] The PR does not introduce [breaking changes](https://github.com/Azure/azure-powershell/blob/preview/documentation/breaking-changes/breaking-changes-definition.md)
- [ ] If applicable, the changes made in the PR have proper test coverage
- [ ] For public API changes to cmdlets:
    - [ ] the changes have gone through a [cmdlet design review](https://github.com/Azure/azure-powershell-cmdlet-review-pr)
    - [ ] the cmdlet markdown files were [generated using the `platyPS` module](https://github.com/Azure/azure-powershell/blob/preview/documentation/development-docs/help-generation.md)

### CODE REVIEW

**Comment by maddieclayton** [comment]
@sergey-shandar Looks like you need to regenerate the WIX installer: https://github.com/Azure/azure-powershell/wiki/Azure-Powershell-Developer-Guide#updating-the-installer

File: src/ResourceManager/Common/Commands.Common.Strategies/DependencyEngine.cs [hunk 1]
```diff
@@ -12,19 +12,20 @@
 // limitations under the License.
 // ----------------------------------------------------------------------------------
 
-using System;
+using System.Collections.Concurrent;
 
 namespace Microsoft.Azure.Commands.Common.Strategies
 {
-    public static class NestedResourceConfigExtensions
+    public sealed class DependencyEngine : IEngine
```
Comment (sergey-shandar) [inline src/ResourceManager/Common/Commands.Common.Strategies/DependencyEngine.cs]: should be internal

File: src/ResourceManager/Common/Commands.Common.Strategies.UnitTest/TargetStateTest.cs [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+        class Model
+        {
+            public string Location { get; set; }
+
+            public IList<NestedModel> Nested { get; set; }
+        }
+
+        [Fact]
+        [Trait(Category.AcceptanceType, Category.CheckIn)]
+        public void Test()
```
Comment (markcowl) [inline src/ResourceManager/Common/Commands.Common.Strategies.UnitTest/TargetStateTest.cs]: This should have a more descriptive method name - reflecting what this is testing (target state defaults?)

File: src/ResourceManager/Common/Commands.Common.Strategies/DependencyEngine.cs (hunk 1 above)
Comment (markcowl) [reply src/ResourceManager/Common/Commands.Common.Strategies/DependencyEngine.cs]: agreed

File: src/ResourceManager/Common/Commands.Common.Strategies/IEngine.cs [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// ----------------------------------------------------------------------------------
+
+namespace Microsoft.Azure.Commands.Common.Strategies
+{
+    public interface IEngine
```
Comment (markcowl) [inline src/ResourceManager/Common/Commands.Common.Strategies/IEngine.cs]: This seems sparse for an interface

**Comment by markcowl** [comment]
on demand run here: https://azuresdkci.westus2.cloudapp.azure.com/view/PowerShell/job/powershell-demand/522/

**Pull request merged**